recipes/CoVoST/AST/train.py

```diff
@@ -0,0 +1,498 @@
+        feats = self.hparams.normalize(feats, wav_lens)
+
+        # Add feature augmentation if specified.
+        if (
+            stage == sb.Stage.TRAIN
+            and hasattr(self.hparams, "fea_augment")
+            and self.optimizer_step > self.hparams.augment_warmup
+        ):
+            feats, fea_lens = self.hparams.fea_augment(feats, wav_lens)
+
+        # forward modules
+        src = self.modules.CNN(feats)
+        enc_out, pred = self.modules.Transformer(
+            src, tokens_bos, wav_lens, pad_idx=self.hparams.pad_index
+        )
+
+        # output layer for ctc ASR log-probabilities if warming up
+        if self.optimizer_step < self.hparams.asr_warmup_steps:
+            logits = self.modules.ctc_lin(enc_out)
+            p_ctc = self.hparams.log_softmax(logits)
+        else:
+            p_ctc = None
+
+        # output layer for seq2seq log-probabilities
+        pred = self.modules.seq_lin(pred)
+        p_seq = self.hparams.log_softmax(pred)
+
+        # Compute outputs
+        hyps = None
+        current_epoch = self.hparams.epoch_counter.current
+        is_valid_search = (
+            stage == sb.Stage.VALID
+            and current_epoch % self.hparams.valid_search_interval == 0
+        )
+        is_test_search = stage == sb.Stage.TEST
+
+        if is_valid_search:
+            hyps, _, _, _ = self.hparams.valid_search(
+                enc_out.detach(), wav_lens
+            )
+
+        elif is_test_search:
+            hyps, _, _, _ = self.hparams.test_search(enc_out.detach(), wav_lens)
+
+        return p_ctc, p_seq, wav_lens, hyps
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC+NLL) given predictions and targets."""
+
+        (p_ctc, p_seq, wav_lens, predicted_tokens) = predictions
+
+        ids = batch.id
+        tokens_eos, tokens_eos_lens = batch.tokens_eos
+        tokens, tokens_lens = batch.tokens
+        tokens_asr, tokens_asr_lens = batch.tokens_asr
+
+        loss_seq = self.hparams.seq_cost(
+            p_seq, tokens_eos, length=tokens_eos_lens
+        )
+
+        # ASR warmup with CTC if specified.
+        if self.optimizer_step < self.hparams.asr_warmup_steps:
+            loss_ctc = self.hparams.ctc_cost(
+                p_ctc, tokens_asr, wav_lens, tokens_asr_lens
+            )
+            loss = (
+                self.hparams.ctc_weight * loss_ctc
+                + (1 - self.hparams.ctc_weight) * loss_seq
+            )
+        else:
+            loss = loss_seq
+
+        if stage != sb.Stage.TRAIN:
+            current_epoch = self.hparams.epoch_counter.current
+            valid_search_interval = self.hparams.valid_search_interval
+            if current_epoch % valid_search_interval == 0 or (
+                stage == sb.Stage.TEST
+            ):
+
+                tgt_detokenizer = MosesDetokenizer(
+                    lang=self.hparams.tgt_language
+                )
```

Comment:
hmmm I think this might require some documentation somewhere. Is there a point of having a wrapper on top of that in speechbrain? Also, I think it would make sense to init this tokenizer outside of this block of code, as every test step you are going to re-init this tokenizer for "nothing". It would add more visibility by defining elsewhere (and only once)